tensorflow-frameworks/tensorflow-data/src/main/java/org/tensorflow/data/Dataset.java

```diff
@@ -0,0 +1,210 @@
+     * Creates an iterator which iterates through all batches of this Dataset in an eager fashion.
+     * Each batch is a list of components, returned as `Output` objects.
+     * <p>
+     * This method enables for-each iteration through batches when running
+     * in eager mode. For Graph mode batch iteration, see `makeOneShotIterator`.
+     *
+     * @return an Iterator through batches of this dataset.
+     */
+    @Override
+    public Iterator<List<Output<?>>> iterator() {
+
+        if (!(tf.scope().env() instanceof EagerSession)) {
+            throw new UnsupportedOperationException("Cannot iterate through a dataset in graph mode.");
+        }
+
+        DatasetIterator iterator = makeOneShotIterator();
+
+        return new Iterator<List<Output<?>>>() {
+            private List<Output<?>> tryNext = getNext();
+
+            private List<Output<?>> getNext() {
+                try {
+                    return iterator.getNext();
+                } catch (IndexOutOfBoundsException e) {
+                    return null;
+                }
+            }
+
+            @Override
+            public boolean hasNext() {
+                return tryNext != null;
+            }
+
+            @Override
+            public List<Output<?>> next() {
+                List<Output<?>> result = tryNext;
+                tryNext = getNext();
+                return result;
+            }
+        };
+    }
+
+    /**
+     * Creates a `DatasetIterator` that can be used to iterate
+     * over elements of this dataset.
+     *
+     * This iterator will have to be initialized with a call
+     * to `iterator.makeInitializer(Dataset)` before elements
+     * can be retreived in a loop.
+     *
+     * @return A new `DatasetIterator` based on this dataset's structure.
+     */
+    public DatasetIterator makeInitializeableIterator() {
+        return DatasetIterator
+                .fromStructure(tf, outputTypes, outputShapes);
+    }
+
+    /**
+     * Creates a `DatasetIterator` that can be used to iterate over
+     * elements of this dataset. Using `makeOneShotIterator` ensures
+     * that the iterator is
+     * automatically initialized on this dataset.
+     *skips
+     * In graph mode, the initializer op will be added to the Graph's
+     * intitializer list, which must be run via `tf.init()`:
+     *
+     * Ex:
+     * <pre>
+     *     try (Session session = new Session(graph) {
+     *         // Immediately run initializers
+     *         session.run(tf.init());
+     *     }
+     * </pre>
+     *
+     * In eager mode, the initializer will be run automatically as a result
+     * of this call.
+     *
+     * @return A new `DatasetIterator` based on this dataset's structure.
+     */
+    public DatasetIterator makeOneShotIterator() {
+        DatasetIterator iterator = makeInitializeableIterator();
+        Op initializer = iterator.makeInitializer(this);
+        if (tf.scope().env() instanceof Graph) tf.initAdd(initializer);

```

Comment:
I think we talked previously to add a method in the `Ops` or `Scope` class telling if we are in graph mode or not without the need of `instanceof` (returning either a boolean or an enum to support future execution environments). Should we do it in this PR as well?

Reply:
I can add this to the `Ops` class! Which file do I change? `tensorflow-core/tensorflow-core-api/gen/annotations/org/tensorflow/op/Ops.java` is generated, right?

Reply:
I added this method to the `ExecutionEnvironment` class; so we can use `tf.scope().env().isEager()`  or `isGraph()` instead.